src/mono/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,167 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Runtime
+{
+    // The ephemeron type is checked in the Mono runtime in "object.c" as follows:
+    //
+    // m_class_get_image (klass) == mono_defaults.corlib && !strcmp ("Ephemeron", m_class_get_name (klass))
+    //
+    // As such, the namespace it belongs to in the managed runtime doesn't matter: the VM will just check
+    // that the type name matches, and that the type is in fact defined in corelib. This means we can
+    // just move it to System.Runtime without worrying about it being properly managed in the VM.
+    // Additionally, the type is defined in "sgen-mono.c" as follows:
+    //
+    // typedef struct {
+    //     GCObject* key;
+    //     GCObject* value;
+    // } Ephemeron;
+    //
+    // So as long as the layout matches the one of the type defined in C# (which it does), we're also free
+    // to rename the fields to better follow the naming guidelines, and the VM will have no issues with it.
+    internal struct Ephemeron
+    {
+        public object? Key;
+        public object? Value;
+    }
+
+    //
+    // Instead of dependent handles, mono uses arrays of Ephemeron objects.
+    //
+    public struct DependentHandle : IDisposable
+    {
+        private Ephemeron[]? _data;
+
+        public DependentHandle(object? target, object? dependent)
+        {
+            _data = new Ephemeron[1];
+            _data[0].Key = target;
+            _data[0].Value = dependent;
+            GC.register_ephemeron_array(_data);
+        }
+
+        public bool IsAllocated => _data is not null;
+
+        public object? Target
+        {
+            get => UnsafeGetTarget();
+            set => UnsafeSetTarget(value);
+        }
+
+        public object? Dependent
+        {
+            get => GetDependent();
```

Comment:
Should this be `UnsafeGetDependent` for consistency?

Reply:
Fixed as part of 64bad6418ff202c60862eabf22468b508dd662f3.